modPrecompiled/src/org/aion/precompiled/ContractFactory.java

```diff
@@ -78,30 +80,59 @@ public IPrecompiledContract getPrecompiledContract(
-                                track,
-                                Address.wrap(ADDR_TOKEN_BRIDGE_INITIAL_OWNER),
-                                Address.wrap(ADDR_TOKEN_BRIDGE));
-
-                if (!context.origin().equals(Address.wrap(ADDR_TOKEN_BRIDGE_INITIAL_OWNER))
-                        && !contract.isInitialized()) {
+        CfgFork cfg = new CfgFork();
+        String forkProperty = cfg.getProperties().getProperty("fork0.3.2");
+        boolean fork_032 =
+                (forkProperty != null) && (context.blockNumber() >= Long.valueOf(forkProperty));
+
+        if (fork_032) {
+            switch (context.address().toString()) {
+                case ADDR_TOKEN_BRIDGE:
+                    TokenBridgeContract contract =
+                            new TokenBridgeContract(
+                                    context,
+                                    track,
+                                    Address.wrap(ADDR_TOKEN_BRIDGE_INITIAL_OWNER),
+                                    Address.wrap(ADDR_TOKEN_BRIDGE));
+
+                    if (!context.origin().equals(Address.wrap(ADDR_TOKEN_BRIDGE_INITIAL_OWNER))
+                            && !contract.isInitialized()) {
+                        return null;
+                    }
+
+                    return contract;
+                case ADDR_ED_VERIFY:
+                    return PC_ED_VERIFY;
+                case ADDR_BLAKE2B_HASH:
+                    return PC_BLAKE2B_HASH;
+                case ADDR_TX_HASH:
+                    return new TXHashContract(context);
+                case ADDR_TOTAL_CURRENCY:
+                default:
                     return null;
-                }
-
-                return contract;
-            case ADDR_ED_VERIFY:
-                return PC_ED_VERIFY;
-            case ADDR_BLAKE2B_HASH:
-                return PC_BLAKE2B_HASH;
-            case ADDR_TX_HASH:
-                return new TXHashContract(context);
-            case ADDR_TOTAL_CURRENCY:
-            default:
-                return null;
+            }
+        } else {
+            switch (context.address().toString()) {
+                case ADDR_TOTAL_CURRENCY:
+                    return new TotalCurrencyContract(
+                            track, context.sender(), Address.wrap(ADDR_OWNER));
+                case ADDR_TOKEN_BRIDGE:
+                    TokenBridgeContract contract =
+                            new TokenBridgeContract(
+                                    context,
+                                    track,
+                                    Address.wrap(ADDR_TOKEN_BRIDGE_INITIAL_OWNER),
+                                    Address.wrap(ADDR_TOKEN_BRIDGE));
+
+                    if (!context.origin().equals(Address.wrap(ADDR_TOKEN_BRIDGE_INITIAL_OWNER))
+                            && !contract.isInitialized()) {
+                        return null;
+                    }
+
+                    return contract;
```

Comment:
the `TokenBridgeContract` code could be moved to a method that can be called by both switch statements